veritone/go-messaging-lib
# VTN-7492 Manual consumer offset commit support

## Overview
Kafka clients sometimes need the ability to only mark the offset as ready for commit when they are done with processing the Kafka message. This commit adds that flexibility and allow clients to mark the offset when they are done processing a particular message.

I also separated the `Consume` method in hope to improve readability.

## Test
* Manual test
I used this library on Coordinator and do:
```bash
consumer, err := kafka.Consumer(topic, consumerGroup, brokers...)
consumer.DisableAutoMark()
```
I ran the service locally and produced some messages to the queue get the message the first time. Then, I exit the program, turned off producer and started again for the program to consume the message once again. I was able to confirm that those messages are re-consumed since those messages were never committed.

To mark the message as processed and ready for auto commit, do on the client:
```
ctx.KafkaChunkConsumer.MarkOffset(msg, "")
```
* Functional test
```bash
VTN-00027-(Duy-Nguyen):kafka duynguyen$ go test -run ^TestConsumerManualCommit
2018/08/17 17:42:51 Broker started: kafka2_1    | [2018-08-18 00:42:51,853] INFO [KafkaServer id=2] started (kafka.server.KafkaServer)
2018/08/17 17:42:51 Broker started: kafka3_1    | [2018-08-18 00:42:51,858] INFO [KafkaServer id=3] started (kafka.server.KafkaServer)
2018/08/17 17:42:51 Broker started: kafka1_1    | [2018-08-18 00:42:51,900] INFO [KafkaServer id=1] started (kafka.server.KafkaServer)
(*kafka.event)(0xc4201be018)({
 Message: (*kafka.Message)(0xc420158930)({
  Value: ([]uint8) (len=5 cap=6) {
   00000000  74 65 73 74 31                                    |test1|
  },
  Key: ([]uint8) (len=4 cap=11) {
   00000000  74 65 73 74                                       |test|
  },
  Offset: (int64) 0,
  Partition: (int32) 0,
  Time: (time.Time) 2018-08-17 17:42:52.295 -0700 PDT,
  Topic: (string) (len=30) "topic_TestConsumerManualCommit"
 })
})
(*kafka.event)(0xc4200b8110)({
 Message: (*kafka.Message)(0xc420316380)({
  Value: ([]uint8) (len=5 cap=6) {
   00000000  74 65 73 74 31                                    |test1|
  },
  Key: ([]uint8) (len=4 cap=11) {
   00000000  74 65 73 74                                       |test|
  },
  Offset: (int64) 0,
  Partition: (int32) 0,
  Time: (time.Time) 2018-08-17 17:42:52.295 -0700 PDT,
  Topic: (string) (len=30) "topic_TestConsumerManualCommit"
 })
})
2018/08/17 17:43:11 Succesfully cleaned up docker-compose docker-compose.kafka.yaml
PASS
ok  	github.com/veritone/go-messaging-lib/kafka	32.685s

```
